fabric-enchantment-api-v1/src/main/java/net/fabricmc/fabric/api/enchantment/v1/EnchantmentEvents.java

```diff
@@ -0,0 +1,126 @@
+ * newer versions of the game.</p>
+ */
+public final class EnchantmentEvents {
+	/**
+	 * Overrides whether or not the given enchantment should be applied to the given item
+	 * only when enchanted using an anvil.
+	 *
+	 * <p>Callbacks are evaluated on a first come, first serve basis where callbacks
+	 * registered earlier will have functional priority over ones registered later.</p>
+	 */
+	public static final Event<AcceptApplication> ACCEPT_APPLICATION = EventFactory.createArrayBacked(AcceptApplication.class, (callbacks) -> (enchantment, stack) -> {
+		for (AcceptApplication callback : callbacks) {
+			TriState result = callback.shouldAccept(enchantment, stack);
+
+			if (result != TriState.DEFAULT) {
+				return result;
+			}
+		}
+
+		return TriState.DEFAULT;
+	});
+
+	/**
+	 * Overrides whether or not the given enchantment should be applied to the given item
+	 * only when enchanted using an enchantment table.
+	 *
+	 * <p>Callbacks are evaluated on a first come, first serve basis where callbacks
+	 * registered earlier will have functional priority over ones registered later.</p>
+	 */
+	public static final Event<AcceptEnchantment> ACCEPT_ENCHANTMENT = EventFactory.createArrayBacked(AcceptEnchantment.class, (callbacks) -> (enchantment, stack) -> {
+		for (AcceptEnchantment callback : callbacks) {
+			TriState result = callback.shouldAccept(enchantment, stack);
+
+			if (result != TriState.DEFAULT) {
+				return result;
+			}
+		}
+
+		return TriState.DEFAULT;
+	});
+
+	/**
+	 * Simple helper method to add an enchantment callback to both the enchantment and application
+	 * events.
+	 *
+	 * @param callback The enchantment callback to add to both enchantment events.
+	 */
+	public static void registerAll(BiFunction<Enchantment, ItemStack, TriState> callback) {
+		ACCEPT_ENCHANTMENT.register((AcceptEnchantment) callback);
+		ACCEPT_APPLICATION.register((AcceptApplication) callback);
+	}
+
+	public interface AcceptEnchantment {
+		/**
+		 * Determines whether or not the given enchantment should be applied to the given item stack
+		 * using an enchantment table.
+		 *
+		 * <p>Returning {@link net.fabricmc.fabric.api.util.TriState#DEFAULT} will delegate to the
+		 * vanilla/modded default functionality for the enchantment. Returning
+		 * {@link net.fabricmc.fabric.api.util.TriState#TRUE} will override the default functionality
+		 * and force the item to accept the enchantment. Returning
+		 * {@link net.fabricmc.fabric.api.util.TriState#FALSE} will override the default functionality
+		 * and prevent the item from accepting the enchantment.</p>
+		 *
+		 * @param enchantment The enchantment looking to be applied.
+		 * @param stack The item stack looking to accept the given enchantment.
+		 * @return Whether or not the given enchantment should be applied to the given item stack
+		 * through the use of an enchantment table.
+		 */
+		TriState shouldAccept(Enchantment enchantment, ItemStack stack);
```

Comment:
It may be worth passing World + nullable BlockPos + nullable Player in these.

Reply:
The interface methods also shadow each other, preventing implementing both differently in the same class.

Reply:
That can definitely be done for the anvil (I did do it for the anvil in dev, but it wasn't possible to get the block's position just the player's position). For the enchantment table that will be significantly more difficult and require actually mixing into the enchantment table and changing how that works (which we could totally do if we want to go down that path, but that would be totally different).